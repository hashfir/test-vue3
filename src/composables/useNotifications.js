import { ref, reactive } from 'vue'

const notifications = ref([])

export function useNotifications() {
  const addNotification = (message, type = 'info', duration = 3000) => {
    const id = Date.now() + Math.random()
    const notification = {
      id,
      message,
      type, // 'success', 'error', 'warning', 'info'
      duration
    }
    
    notifications.value.push(notification)
    
    if (duration > 0) {
      setTimeout(() => {
        removeNotification(id)
      }, duration)
    }
    
    return id
  }
  
  const removeNotification = (id) => {
    const index = notifications.value.findIndex(n => n.id === id)
    if (index > -1) {
      notifications.value.splice(index, 1)
    }
  }
  
  const showSuccess = (message) => addNotification(message, 'success')
  const showError = (message) => addNotification(message, 'error')
  const showWarning = (message) => addNotification(message, 'warning')
  const showInfo = (message) => addNotification(message, 'info')
  
  const clearAll = () => {
    notifications.value = []
  }
  
  return {
    notifications,
    addNotification,
    removeNotification,
    showSuccess,
    showError,
    showWarning,
    showInfo,
    clearAll
  }
}