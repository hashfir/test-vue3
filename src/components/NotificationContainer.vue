<template>
  <div class="notification-container">
    <transition-group name="notification" tag="div">
      <div
        v-for="notification in notifications"
        :key="notification.id"
        :class="['notification', `notification--${notification.type}`]"
      >
        <div class="notification__content">
          {{ notification.message }}
        </div>
        <button 
          class="notification__close"
          @click="removeNotification(notification.id)"
        >
          Ã—
        </button>
      </div>
    </transition-group>
  </div>
</template>

<script setup>
import { useNotifications } from '../composables/useNotifications'

const { notifications, removeNotification } = useNotifications()
</script>

<style scoped>
.notification-container {
  position: fixed;
  top: 20px;
  right: 20px;
  z-index: 1000;
  max-width: 400px;
}

.notification {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 12px 16px;
  margin-bottom: 8px;
  border-radius: 6px;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.1);
  background: white;
  border-left: 4px solid #ccc;
}

.notification--success {
  border-left-color: #10b981;
  background-color: #f0fdf4;
  color: #065f46;
}

.notification--error {
  border-left-color: #ef4444;
  background-color: #fef2f2;
  color: #991b1b;
}

.notification--warning {
  border-left-color: #f59e0b;
  background-color: #fffbeb;
  color: #92400e;
}

.notification--info {
  border-left-color: #3b82f6;
  background-color: #eff6ff;
  color: #1e40af;
}

.notification__content {
  flex: 1;
  font-size: 14px;
  line-height: 1.4;
}

.notification__close {
  background: none;
  border: none;
  font-size: 20px;
  font-weight: bold;
  cursor: pointer;
  color: inherit;
  opacity: 0.7;
  margin-left: 12px;
  padding: 0;
  line-height: 1;
}

.notification__close:hover {
  opacity: 1;
}

.notification-enter-active,
.notification-leave-active {
  transition: all 0.3s ease;
}

.notification-enter-from {
  opacity: 0;
  transform: translateX(100%);
}

.notification-leave-to {
  opacity: 0;
  transform: translateX(100%);
}
</style>