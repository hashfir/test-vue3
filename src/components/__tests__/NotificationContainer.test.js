import { describe, it, expect, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import NotificationContainer from '../NotificationContainer.vue'
import { useNotifications } from '../../composables/useNotifications'

describe('NotificationContainer', () => {
  let wrapper
  let notifications

  beforeEach(() => {
    wrapper = mount(NotificationContainer)
    notifications = useNotifications()
    // Clear any existing notifications
    notifications.clearAll()
  })

  it('should render empty when no notifications', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render success notification', async () => {
    notifications.showSuccess('Test success message')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render error notification', async () => {
    notifications.showError('Test error message')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render multiple notifications', async () => {
    notifications.showSuccess('Success message')
    notifications.showError('Error message')
    notifications.showWarning('Warning message')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should remove notification when close button is clicked', async () => {
    notifications.showSuccess('Test message')
    await wrapper.vm.$nextTick()
    
    const closeButton = wrapper.find('.notification__close')
    expect(closeButton.exists()).toBe(true)
    
    await closeButton.trigger('click')
    await wrapper.vm.$nextTick()
    
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should apply correct CSS classes for different notification types', async () => {
    notifications.showSuccess('Success')
    notifications.showError('Error')
    notifications.showWarning('Warning')
    notifications.showInfo('Info')
    await wrapper.vm.$nextTick()

    expect(wrapper.find('.notification--success').exists()).toBe(true)
    expect(wrapper.find('.notification--error').exists()).toBe(true)
    expect(wrapper.find('.notification--warning').exists()).toBe(true)
    expect(wrapper.find('.notification--info').exists()).toBe(true)
  })
})