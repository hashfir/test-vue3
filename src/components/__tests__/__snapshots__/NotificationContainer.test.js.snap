// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationContainer > should remove notification when close button is clicked 1`] = `
"<div data-v-8470eda0="" class="notification-container">
  <transition-group-stub data-v-8470eda0="" name="notification" tag="div" appear="false" persisted="false" css="true"></transition-group-stub>
</div>"
`;

exports[`NotificationContainer > should render empty when no notifications 1`] = `
"<div data-v-8470eda0="" class="notification-container">
  <transition-group-stub data-v-8470eda0="" name="notification" tag="div" appear="false" persisted="false" css="true"></transition-group-stub>
</div>"
`;

exports[`NotificationContainer > should render error notification 1`] = `
"<div data-v-8470eda0="" class="notification-container">
  <transition-group-stub data-v-8470eda0="" name="notification" tag="div" appear="false" persisted="false" css="true">
    <div data-v-8470eda0="" class="notification notification--error">
      <div data-v-8470eda0="" class="notification__content">Test error message</div><button data-v-8470eda0="" class="notification__close"> × </button>
    </div>
  </transition-group-stub>
</div>"
`;

exports[`NotificationContainer > should render multiple notifications 1`] = `
"<div data-v-8470eda0="" class="notification-container">
  <transition-group-stub data-v-8470eda0="" name="notification" tag="div" appear="false" persisted="false" css="true">
    <div data-v-8470eda0="" class="notification notification--success">
      <div data-v-8470eda0="" class="notification__content">Success message</div><button data-v-8470eda0="" class="notification__close"> × </button>
    </div>
    <div data-v-8470eda0="" class="notification notification--error">
      <div data-v-8470eda0="" class="notification__content">Error message</div><button data-v-8470eda0="" class="notification__close"> × </button>
    </div>
    <div data-v-8470eda0="" class="notification notification--warning">
      <div data-v-8470eda0="" class="notification__content">Warning message</div><button data-v-8470eda0="" class="notification__close"> × </button>
    </div>
  </transition-group-stub>
</div>"
`;

exports[`NotificationContainer > should render success notification 1`] = `
"<div data-v-8470eda0="" class="notification-container">
  <transition-group-stub data-v-8470eda0="" name="notification" tag="div" appear="false" persisted="false" css="true">
    <div data-v-8470eda0="" class="notification notification--success">
      <div data-v-8470eda0="" class="notification__content">Test success message</div><button data-v-8470eda0="" class="notification__close"> × </button>
    </div>
  </transition-group-stub>
</div>"
`;
